src/Lecture3.hs

```diff
@@ -159,10 +172,15 @@ monsters, you should get a combined treasure and not just the first
 🕯 HINT: You may need to add additional constraints to this instance
   declaration.
 -}
-instance Semigroup (Treasure a) where
+instance (Semigroup a) => Semigroup (Treasure a) where
+  NoTreasure <> NoTreasure = NoTreasure
+  (SomeTreasure t1) <> NoTreasure = SomeTreasure t1
+  NoTreasure <> (SomeTreasure t1) = SomeTreasure t1
```

Comment:
This is an absolutely correct solution!

However, we can notice that these 3 cases could be reduced to 2 with simpler pattern matching!
Let's see what we can unite in here.

If some of the arguments are `NoTreasure`, you actually always return the second one! So you need to inspect only two cases: `NoTreasure` in the first position, and `NoTreasure` in the second one 🙂